Adds Quirk so that booking.com sends simple JPEG to WebKit on iOS

karlcow opened the pull request:
#### 7c18c3da4148c58fdb27ede24514d32ab6bd3ab6
<pre>
Adds Quirk so that booking.com sends simple JPEG to WebKit on iOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=269875">https://bugs.webkit.org/show_bug.cgi?id=269875</a>
<a href="https://rdar.apple.com/123408770">rdar://123408770</a>

Reviewed by Chris Dumez.

In webkit.org/b/178758, the support for JPEG2000 was removed from
WebKit. Booking.com is doing content negociation on both the
User-Agent string and the HTTP accept header. The website sends
JPEG2000 assets to Safari on iOS. An outreach effort has been started
to make sure that booking fixes their servers.
See <a href="https://github.com/webcompat/web-bugs/issues/133697">https://github.com/webcompat/web-bugs/issues/133697</a>

No other breakage has been detected yet, but we know that Cloudinary
is currently JPEG2000 images for some websites on both iOS and macOS.
It might break once this release of WebKit reaches Safari. An ongoing
effort has started to avoid this so that Cloudinary can adjust their
rules and warn their customers.

In the meantime this Quirk sends a different HTTP Accept header for
booking.com tricking the server into sending compatible image resources
for HTML requests instead of JPEG2000.

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::updateRequestAndAddExtraFields):
* Source/WebCore/loader/cache/CachedResourceRequest.cpp:
(WebCore::CachedResourceRequest::acceptHeaderValueFromType):
(WebCore::CachedResourceRequest::setAcceptHeaderIfNone):
* Source/WebCore/loader/cache/CachedResourceRequest.h:
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldSendLongerAcceptHeaderQuirk):
* Source/WebCore/page/Quirks.h:

Canonical link: <a href="https://commits.webkit.org/275216@main">https://commits.webkit.org/275216@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eb62249d79557fcd33df91a80f3708ba8874a1cb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41161 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20174 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43539 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43724 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37253 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43468 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23216 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17505 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34066 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41735 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17102 "Found 1 new test failure: http/wpt/background-fetch/background-fetch-persistency.window.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35456 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14711 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14841 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36457 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45044 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37355 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36772 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40515 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15976 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13101 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38893 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17595 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17647 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5497 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/17239 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -3212,7 +3212,7 @@ void FrameLoader::updateRequestAndAddExtraFields(ResourceRequest& request, IsMai
 
     if (isMainResource)
-        request.setHTTPHeaderField(HTTPHeaderName::Accept, CachedResourceRequest::acceptHeaderValueFromType(CachedResource::Type::MainResource));
+        request.setHTTPHeaderField(HTTPHeaderName::Accept, CachedResourceRequest::acceptHeaderValueFromType(CachedResource::Type::MainResource, request));
```
I think we should also pass `protectedFrame().ptr()` here.

cdumez commented on Source/WebCore/loader/cache/CachedResourceRequest.cpp:
```diff
@@ -190,7 +194,7 @@ String CachedResourceRequest::acceptHeaderValueFromType(CachedResource::Type typ
 {
     if (!m_resourceRequest.hasHTTPHeader(HTTPHeaderName::Accept))
-        m_resourceRequest.setHTTPHeaderField(HTTPHeaderName::Accept, acceptHeaderValueFromType(type));
+        m_resourceRequest.setHTTPHeaderField(HTTPHeaderName::Accept, acceptHeaderValueFromType(type, m_resourceRequest));
```
Then also pass the frame here.

cdumez commented on Source/WebCore/loader/cache/CachedResourceRequest.cpp:
```diff
@@ -190,7 +194,7 @@ String CachedResourceRequest::acceptHeaderValueFromType(CachedResource::Type typ
 void CachedResourceRequest::setAcceptHeaderIfNone(CachedResource::Type type)
```
We should also take in a `LocalFrame*`, there is a simple call site and this call site call pass `protectedFrame().get()`

cdumez commented on Source/WebCore/loader/cache/CachedResourceRequest.cpp:
```diff
@@ -166,10 +166,14 @@ static String acceptHeaderValueForImageResource()
 }
 
-String CachedResourceRequest::acceptHeaderValueFromType(CachedResource::Type type)
+String CachedResourceRequest::acceptHeaderValueFromType(CachedResource::Type type, ResourceRequest& request)
```
Should take in a `LocalFrame*` too

karlcow replied:
* `acceptHeaderValueFromType(type, m_resourceRequest, frame)` is called from `CachedResourceRequest::setAcceptHeaderIfNone(CachedResource::Type type, LocalFrame* frame)`
* Then `request.setAcceptHeaderIfNone(type, frame)` is called from `CachedResourceLoader::prepareFetch(CachedResource::Type type, CachedResourceRequest& request, LocalFrame* frame)`
* Then `prepareFetch(type, request, frame)` is called from `ResourceErrorOr<CachedResourceHandle<CachedResource>> CachedResourceLoader::requestResource(CachedResource::Type type, CachedResourceRequest&& request, ForPreload forPreload, ImageLoading imageLoading)`

which gets me 

```
The following build commands failed:
	CompileC /Users/karlcow/code/OpenSource/WebKitBuild/WebCore.build/Debug/WebCore.build/Objects-normal/arm64e/UnifiedSource256.o /Users/karlcow/code/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource256.cpp normal arm64e c++ com.apple.compilers.llvm.clang.1_0.compiler (in target 'WebCore' from project 'WebCore')
In file included from /Users/karlcow/code/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource256.cpp:6:
(1 failure)
/Users/karlcow/code/OpenSource/Source/WebCore/loader/cache/CachedResourceLoader.cpp:1036:33: error: no viable conversion from 'Ref<LocalFrame>' (aka 'WTF::Ref<WebCore::LocalFrame>') to 'LocalFrame *'
    prepareFetch(type, request, frame);
In file included from /Users/karlcow/code/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource256.cpp:1:
In file included from /Users/karlcow/code/OpenSource/Source/WebCore/WebCorePrefix.h:165:
In file included from /Users/karlcow/code/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/HashMap.h:25:
In file included from /Users/karlcow/code/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/HashTable.h:33:
In file included from /Users/karlcow/code/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/HashTraits.h:26:
In file included from /Users/karlcow/code/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/HashFunctions.h:26:
In file included from /Users/karlcow/code/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/RefPtr.h:28:
/Users/karlcow/code/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/Ref.h:136:5: note: candidate function
    operator T&() const { ASSERT(m_ptr); return *PtrTraits::unwrap(m_ptr); }
In file included from /Users/karlcow/code/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource256.cpp:6:
/Users/karlcow/code/OpenSource/Source/WebCore/loader/cache/CachedResourceLoader.cpp:902:112: note: passing argument to parameter 'frame' here
void CachedResourceLoader::prepareFetch(CachedResource::Type type, CachedResourceRequest& request, LocalFrame* frame)
1 error generated.
make: *** [d] Error 65
```

cdumez replied:
I already said in my earlier review but you don't need to change prepareFetch() to take in a frame. `CachedResourceLoader::prepareFetch()` can just pass `protectedFrame().get()` to `request.setAcceptHeaderIfNone()`.

cdumez replied:
`CachedResourceLoader::protectedFrame()` already exists.

karlcow replied:
Ha understood now. Sorry about that! Thanks a lot.

cdumez commented on Source/WebCore/loader/cache/CachedResourceRequest.cpp:
```diff
@@ -166,10 +166,14 @@ static String acceptHeaderValueForImageResource()
+
     switch (type) {
     case CachedResource::Type::MainResource:
+        if (Quirks::shouldSendLongerAcceptHeaderQuirk(host))
```
I'd pass in the frame here.

cdumez commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1716,6 +1716,16 @@ bool Quirks::shouldDisableNavigatorStandaloneQuirk() const
 
+// booking.com https://webkit.org/b/269875
+// FIXME: booking.com https://webkit.org/b/269876 when outreach has been successful.
+bool Quirks::shouldSendLongerAcceptHeaderQuirk(StringView host)
```
We should take in a `LocalFrame*` here too.

cdumez replied:
Then add this check:
```
if (frame && !frame->settings().needsSiteSpecificQuirks())
    return false;
```

cdumez reviewed (changes requested):
This is pretty good but I think it'd be nicer if the developer could disable the quirk to validate any change they make on their side. I made suggestion to try and account for the `disableQuirks` setting when a LocalFrame is available.

cdumez commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1716,6 +1716,19 @@ bool Quirks::shouldDisableNavigatorStandaloneQuirk() const
+    if (frame && !frame->settings().needsSiteSpecificQuirks())
+        return false;
+
+    if (equalLettersIgnoringASCIICase(host, "www.booking.com"_s))
```
If I search for a flight, I'll end up with a URL like "https://flights.booking.com", which is no longer going to match. Isn't this an issue? Is there a reason we only want to do this for "www.booking.com" and not "*.booking.com"?

karlcow replied:
surprisingly flights.booking.com doesn't send JPEG2000 images.

cdumez replied:
So we're confident taking care of "www.booking.com" is sufficient and we don't have to worry about other subdomains (flights was just one example).

karlcow replied:
also flights.booking.com seems to redirect to www.booking.com at a point

cdumez replied:
booking.com seems to have a lot of subdomains:
https://www.vedbex.com/subdomain-finder/booking.com

karlcow replied:
Modified the PR! Thanks.

cdumez commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1716,6 +1716,19 @@ bool Quirks::shouldDisableNavigatorStandaloneQuirk() const
+    if (frame && !frame->settings().needsSiteSpecificQuirks())
+        return false;
+
+    if (equalLettersIgnoringASCIICase(host, "booking.com"_s) || host.endsWithIgnoringASCIICase(".booking.com"_s))
```
We don't usually need to ignore case for host checks. The URL class makes sure the host is lowercase already. For this reason also, it'd be nicer if the function took in a URL instead of a StringView.

cdumez commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1716,6 +1716,21 @@ bool Quirks::shouldDisableNavigatorStandaloneQuirk() const
+// FIXME: booking.com https://webkit.org/b/269876 when outreach has been successful.
+bool Quirks::shouldSendLongerAcceptHeaderQuirk(const URL& url, LocalFrame* frame)
+{
+    auto host = url.host();
```
Please move this *after* the early return. No point in doing this if we're about to early return (common case).

cdumez commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1716,6 +1716,21 @@ bool Quirks::shouldDisableNavigatorStandaloneQuirk() const
+    if (frame && !frame->settings().needsSiteSpecificQuirks())
+        return false;
+
+    if (equalLettersIgnoringASCIICase(host, "booking.com"_s) || host.endsWithIgnoringASCIICase(".booking.com"_s))
```
This still ignores ASCII case.

cdumez reviewed (approved):
LGTM

karlcow committed:
Adds Quirk so that booking.com sends simple JPEG to WebKit on iOS
https://bugs.webkit.org/show_bug.cgi?id=269875
rdar://123408770

Reviewed by Chris Dumez.

In webkit.org/b/178758, the support for JPEG2000 was removed from
WebKit. Booking.com is doing content negociation on both the
User-Agent string and the HTTP accept header. The website sends
JPEG2000 assets to Safari on iOS. An outreach effort has been started
to make sure that booking fixes their servers.
See https://github.com/webcompat/web-bugs/issues/133697

No other breakage has been detected yet, but we know that Cloudinary
is currently JPEG2000 images for some websites on both iOS and macOS.
It might break once this release of WebKit reaches Safari. An ongoing
effort has started to avoid this so that Cloudinary can adjust their
rules and warn their customers.

In the meantime this Quirk sends a different HTTP Accept header for
booking.com tricking the server into sending compatible image resources
for HTML requests instead of JPEG2000.

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::updateRequestAndAddExtraFields):
* Source/WebCore/loader/cache/CachedResourceRequest.cpp:
(WebCore::CachedResourceRequest::acceptHeaderValueFromType):
(WebCore::CachedResourceRequest::setAcceptHeaderIfNone):
* Source/WebCore/loader/cache/CachedResourceRequest.h:
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldSendLongerAcceptHeaderQuirk):
* Source/WebCore/page/Quirks.h:

Canonical link: https://commits.webkit.org/275216@main
